### PR TITLE
Ignite Rework

### PR DESCRIPTION
**Pull request by pollend**
This is based off of: https://github.com/micronaut-projects/micronaut-ignite/pull/29 I'm going to split out all the data stuff for later since that will take more work.

### CODE REVIEW

**Comment by pollend** [comment]
Here is the relevant discussion for some context: http://apache-ignite-developers.2346864.n4.nabble.com/Re-Micronaut-Ignite-td48810.html

**Comment by pollend** [comment]
@dmagda  asked me about the version of ignite that would be packaged with the release.

```
Michael,

The docs look good enough for the release. Once the integration is
released, I'll update and extend my Micronaut tutorial elucidating on how
to use the integration in practice for various use cases.

Btw, are the ignite-core and ignite-cache Maven artifacts will be released
for a specific version of Ignite? Ideally, the version of Ignite needs to
be set in a pom.xml and the Micronaut integration should pick them up.

Also, what are we doing next - reviewing APIs of the ignite-cache module or
going through a review process of the ignite-core and merging it first to
the Micronaut repository?

```

Could the CI be setup to accommodate this?

**Comment by pollend** [comment]
Here is a sample application I put together to show the configuration: https://github.com/pollend/micronaut-ignite-sample

**Comment by graemerocher** [comment]
so @pollend is this ready for you review from your perspective?

**Comment by pollend** [comment]
I went through and excluded items from the configuration that can't be set. This should clean up the documentation. Something I noticed that looks like a bug is `ignite.include-event-types`.

image
it should be `int[]`

**Comment by graemerocher** [comment]
@dmagda are you good with this PR?

File: ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * wraps a call in an ignite transaction.
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Around
+@Type(IgniteTransactionInterceptor.class)
+public @interface IgniteTransaction {
```
Comment (dmagda) [inline ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java:36]: Another thought, do we really need this interface? Application developers can get access to the streaming transactional capabilities of Ignite by injecting an instance of `Ignite` (created by your integration) into their code and by using the `ignite.transactions(..)` method next.

**Review by dmagda — COMMENTED** [review]
@pollend please have a look at my suggestions, overall, great job! Let's polish it up a little bit.

**Comment by dmagda** [comment]
> I went through and excluded items from the configuration that can't be set. This should clean up the documentation.

Btw, why the `description` column of the configuration items is empty on the documentation end? Is there any way to get it populated? Otherwise, application developers have no idea what a configuration option does.

**Comment by pollend** [comment]
I was looking at micronaut-gcp and it looks like the descriptions are filled out manually. not really sure how the descriptions would be filled out.  

https://github.com/micronaut-projects/micronaut-gcp/blob/master/src/main/docs/guide/pubsub/producer/publisherProperties.adoc

File: ignite-core/src/main/java/io/micronaut/ignite/intercept/IgniteTransactionInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import javax.inject.Singleton;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Interceptor implementation for the {@link IgniteTransaction} annotation.
+ */
+@Singleton
+public class IgniteTransactionInterceptor implements MethodInterceptor<Object, Object> {
```
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-core/src/main/java/io/micronaut/ignite/intercept/IgniteTransactionInterceptor.java:43]: I did this a while ago. the basic gist was to allow for ignite transactions by wrapping a method with aop. This is what I gathered from the documentation, might not be relevant. 

```
@IgniteTransaction()
void method(){
 ... cache operations
}
```

File: ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * ignite client by id. uses primary client by default.
+     * @return ignite client id
+     */
+    String client() default "default";
```
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java:42]: ignite is already inferred wouldn't instance() be better?

File: ignite-cache/src/main/java/io/micronaut/ignite/IgniteCacheManager.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import javax.inject.Singleton;
+import java.util.concurrent.ExecutorService;
+
+@Singleton
+@Requires(beans = Ignite.class)
+@Requires(property = DefaultIgniteConfiguration.PREFIX + "." + "enabled", value = StringUtils.TRUE, defaultValue = StringUtils.FALSE)
+public class IgniteCacheManager implements DynamicCacheManager<IgniteCache> {
+    private final Ignite ignite;
+    private final ConversionService<?> service;
+    private final ExecutorService executorService;
```
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-cache/src/main/java/io/micronaut/ignite/IgniteCacheManager.java:40]: I pass it into listenAsync. in micronaut you can configure the executor for managing threads. I'm not sure if its really necessary from micronaut or ignite.  

https://docs.micronaut.io/latest/guide/index.html#io.micronaut.scheduling.executor.UserExecutorConfiguration

```
 igniteFuture.listenAsync((IgniteInClosure<IgniteFuture<Object>>) response -> {
            try {
                Object result = response.get();
                newFuture.complete(conversionService.convert(result, ConversionContext.of(requiredType)));
            } catch (IgniteException ex) {
                newFuture.completeExceptionally(ex);
            }
        }, executorService);
```

File: ignite-core/src/main/java/io/micronaut/ignite/intercept/IgniteTransactionInterceptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmagda) [reply ignite-core/src/main/java/io/micronaut/ignite/intercept/IgniteTransactionInterceptor.java:43]: Got it! Yes, it makes sense to support this feature. I would suggest putting an example or mentioning in the documentation that this feature is available through the integration.

File: ignite-cache/src/main/java/io/micronaut/ignite/IgniteCacheManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmagda) [reply ignite-cache/src/main/java/io/micronaut/ignite/IgniteCacheManager.java:40]: Now I see. Didn't spot the place where the executorService was passed in the function. Alright, let's leave it as is, makes sense.

File: ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmagda) [reply ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java]: How about `instanceName?`

**Comment by dmagda** [comment]
> I was looking at micronaut-gcp and it looks like the descriptions are filled out manually. not really sure how the descriptions are filled out from the annotation process. something similar would have to be arranged right?
> 
> https://github.com/micronaut-projects/micronaut-gcp/blob/master/src/main/docs/guide/pubsub/producer/publisherProperties.adoc

Could you populate this table manually by copy-pasting from the Ignite JavaDocs? Probably, the Micronaut folks might suggest some automatic approach here.

File: ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java]: @Name qualifier is different then the name that is used in the instanceName for the configuration right?

File: ignite-core/src/main/java/io/micronaut/ignite/IgniteFactory.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+     * @return create ignite instance
+     */
+    @Singleton
+    @EachBean(IgniteConfiguration.class)
+    @Bean(preDestroy = "close")
+    public Ignite ignite(IgniteConfiguration configuration) {
+        try {
+            return Ignition.start(configuration);
+        } catch (Exception e) {
+            LOG.error("Failed to instantiate Ignite Client: " + e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-core/src/main/java/io/micronaut/ignite/IgniteFactory.java]: I updated the description.

File: ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmagda) [reply ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java]: You can try out this trick:

- Capture the name used for an Ignite instance in the Micronaut config file (you use the "default" by default)
- Pass this name into the IgniteConfiguration.instanceName before starting the instance
- The ambiguity is solved! Right?

File: ignite-core/src/main/java/io/micronaut/ignite/intercept/IgniteTransactionInterceptor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-core/src/main/java/io/micronaut/ignite/intercept/IgniteTransactionInterceptor.java:43]: I'll add that to the documentation.

File: ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-core/src/main/java/io/micronaut/ignite/annotation/IgniteTransaction.java]: hmm, that sounds good. I'll do that

File: ignite-cache/src/main/java/io/micronaut/ignite/IgniteSyncCache.java [hunk 6]
```diff
@@ -84,7 +84,7 @@ public IgniteSyncCache(ConversionService<?> conversionService, IgniteCache nativ
     public void put(@NonNull Object key, @NonNull Object value) {
         ArgumentUtils.requireNonNull("key", key);
         ArgumentUtils.requireNonNull("value", value);
-        nativeCache.putIfAbsent(key, value);
+        nativeCache.put(key, value);
```
[the remark this answers is not in the document]
Comment (pollend) [reply ignite-cache/src/main/java/io/micronaut/ignite/IgniteSyncCache.java:87]: I believe so

**Comment by pollend** [comment]
I'm looking at this and that's quite a lot of descriptions to copy in manually. my worry is having the configuration go out of sync with the documentation.

**Comment by ilopmar** [comment]
Regarding the descriptions of the documentation not being populated, this is how it works:
- We have these dependency https://github.com/micronaut-projects/micronaut-ignite/blob/bce805493c1abee4be2e2c89749d1f22189b5a23/ignite-core/build.gradle#L4 that extracts the configuration properties and generated the appropriate .adoc files
- The you can include that file in the documentation: https://raw.githubusercontent.com/micronaut-projects/micronaut-ignite/bce805493c1abee4be2e2c89749d1f22189b5a23/src/main/docs/guide/igniteCache/igniteClientCache.adoc

Our `micronaut-docs-asciidoc-config-props` library extracts the descriptions from the source code during the compilation of the module itself but it doesn't work for other external dependencies (because the code has already been compiled). In all the documentation for the rest of Micronaut modules the behavior is the same.

Regarding the GCP documentation, those tables with the descriptions were moved to source control and updated manually. Although I see the point of having the description there I'm not completely sure it's the best approach because those tables need to be maintained manually.


Finally, regarding the PR itself, is it ready for merge now?

**Comment by dmagda** [comment]
@pollend, the changes look good to me now. Please go ahead and merge unless the Micronaut community suggests any other improvements. All is neat and clean from the Ignite standpoint.

**Review by ilopmar — APPROVED** [review]

**Comment by graemerocher** [comment]
Thanks for the contribution!

**Comment by graemerocher** [comment]
@pollend @dmagda would you like to do a 1.0.0.RC1 release with the current state of the code?

**Comment by ilopmar** [comment]
Hold on, I'm going to fix the 2 open issues about the docs, cleanup a few project-config things and them we can release RC1

**Comment by ilopmar** [comment]
We're ready for the release whenever you want. 

I've also submitted this draft PR to core to add this to the bom: https://github.com/micronaut-projects/micronaut-core/pull/4070

**Pull request merged**